studio/src/aem/content-navigation.js

```diff
@@ -64,72 +77,131 @@ class ContentNavigation extends LitElement {
         this.unregisterFromSource();
     }
 
+    toggleFilterPanel() {
+        this.showFilterPanel = !this.showFilterPanel;
+    }
+
+    handlerSourceLoad() {
+        if (this.#initFromFragmentId) {
+            this.#initialFolder = getTopFolder(this.source.fragments[0]?.path);
+            this.fragmentFromIdLoaded = true;
+        }
+        this.forceUpdate();
+    }
+
     registerToSource() {
         this.source = document.getElementById(this.getAttribute('source'));
         if (!this.source) return;
-        this.source.addEventListener(EVENT_LOAD, this.forceUpdate);
+        this.deeplinkDisposer = deeplink(({ path, query }) => {
+            this.#initialFolder =
+                path !== '/content/dam/mas' ? path?.split('/')?.pop() : null;
+            if (!this.#initialFolder && this.source.isFragmentId(query)) {
+                document.querySelector('mas-studio').searchText = query;
+                this.source.searchFragments();
+                this.#initFromFragmentId = true;
+            }
+        });
+        this.source.addEventListener(
+            EVENT_LOAD,
+            this.handlerSourceLoad.bind(this),
+        );
         this.source.addEventListener(EVENT_CHANGE, this.forceUpdate);
+        this.source.getTopFolders().then((folders) => {
+            this.topFolders = folders;
+        });
     }
 
     async forceUpdate() {
         this.requestUpdate();
     }
 
     unregisterFromSource() {
-        this.source?.removeEventListener(EVENT_LOAD, this.forceUpdate);
+        if (this.deeplinkDisposer) {
+            this.deeplinkDisposer();
+        }
+        this.source?.removeEventListener(EVENT_LOAD, this.handlerSourceLoad());
```

Comment:
` this.source?.removeEventListener(EVENT_LOAD, this.handlerSourceLoad());` will not remove the listener.

I recommend to bind the handler in the constructor and then, add and remove it as follows:
```js
this.source?.addEventListener(EVENT_LOAD, this.handlerSourceLoad);
this.source?.removeEventListener(EVENT_LOAD, this.handlerSourceLoad);
```

Reply:
so that the same instance is added and removed.